tests/testsuite/standard_lib.rs

```diff
@@ -0,0 +1,326 @@
+        execs.arg("--target").arg(rustc_host());
+    }
+    execs
+        .arg(unstable)
+        .arg("-Zno-index-update")
+        .env_remove("CARGO_HOME")
+        .env_remove("HOME")
+        .env("CARGO_TARGET_DIR", target.as_os_str())
+        .masquerade_as_nightly_cargo();
+    execs
+}
+
+#[cargo_test]
+fn std_lib() {
+    if !is_nightly() {
+        // -Zbuild-std is nightly
+        // -Zno-index-update is nightly
+        // We don't want these tests to run on rust-lang/rust.
+        return;
+    }
+    simple_lib_std();
+    simple_bin_std();
+    lib_nostd();
+    bin_nostd();
+    check_core();
+    cross_custom();
+    hashbrown();
+    libc();
+    test();
+    target_proc_macro();
+    bench();
+    doc();
+    check_std();
+    doctest();
+}
+
+fn simple_lib_std() {
+    let p = project().file("src/lib.rs", "").build();
+    cargo_build_std(&p, "build -v", "")
+        .with_stderr_contains("[RUNNING] `rustc [..]--crate-name std [..]")
+        .run();
+    // Check freshness.
+    p.change_file("src/lib.rs", " ");
+    cargo_build_std(&p, "build -v", "std")
+        .with_stderr_contains("[FRESH] std[..]")
+        .run();
+}
+
+fn simple_bin_std() {
+    let p = project().file("src/main.rs", "fn main() {}").build();
+    cargo_build_std(&p, "run -v", "std").run();
+}
+
+fn lib_nostd() {
+    let p = project()
+        .file(
+            "src/lib.rs",
+            r#"
+            #![no_std]
+            pub fn foo() {
+                assert_eq!(core::u8::MIN, 0);
+            }
+            "#,
+        )
+        .build();
+    cargo_build_std(&p, "build -v --lib", "core")
+        .with_stderr_does_not_contain("[..]libstd[..]")
+        .run();
+}
+
+fn bin_nostd() {
+    if cfg!(windows) {
+        // I think windows requires setting up mainCRTStartup,
+        // I'm not in the mood to figure it out.
+        return;
+    }
+    let p = project()
+        .file("src/lib.rs", "#![no_std] pub fn foo() {}")
+        .file(
+            "src/main.rs",
+            r#"
+            #![no_std]
+            #![feature(lang_items, start, core_intrinsics)]
+
+            use core::panic::PanicInfo;
+
+            #[panic_handler]
+            fn panic(_info: &PanicInfo) -> ! {
+                unsafe { core::intrinsics::abort() }
+            }
+
+            #[start]
+            fn start(_argc: isize, _argv: *const *const u8) -> isize {
+                foo::foo();
+                123
+            }
+
+            #[lang = "eh_personality"]
+            extern fn eh_personality() {}
+            "#,
+        )
+        .file(
+            "build.rs",
+            r#"
+            fn main() {
+                let target = std::env::var("TARGET").expect("TARGET was not set");
+                if target.contains("apple-darwin") {
+                    println!("cargo:rustc-link-lib=System");
+                } else if target.contains("linux") {
+                    // TODO: why is this needed?
```

Comment:
FWIW I wouldn't bother with a test like this. Producing a non-`std` executable that actually works on all platforms is stupidly complicated and unstable. We used to have a test in rust-lang/rust called `smallest-hello-world.rs` and the [history of that file is brutal](https://github.com/rust-lang/rust/commits/master/src/test/run-pass/smallest-hello-world.rs) (the last commit removed the test).

Reply:
Ah, yea, fair enough.  I'll remove it, though it was a good exercise.